### ISSUE TITLE
Allow accessing protected and private attributes in __new__

### ISSUE THREAD

**Issue by exhuma** [PyCQA/pylint#1802]
When implementing `__new__`, you usually create an instance of the class in which the `__new__` is defined. `pylint` currently does not allow setting protected and private attributes inside `__new__`.

### Steps to reproduce
Lint the follwing code

```python
# pylint: disable=blacklisted-name, missing-docstring

class MyObj:

    def __new__(cls, foo, bar):
        obj = object.__new__(cls)
        obj._foo = foo   # <-- warning reported here
        obj.__bar = bar   # <-- warning reported here
        return obj

    @property
    def foo(self):
        return self._foo

    @property
    def bar(self):
        return self.__bar



def main():
    instance = MyObj(1, 2)
    print(instance.foo)
    print(instance.bar)


main()
```

### Current behavior

This currently returns the following:

```
************* Module example
W:  7, 8: Access to a protected member _foo of a client class (protected-access)
W:  8, 8: Access to a protected member __bar of a client class (protected-access)

------------------------------------------------------------------
Your code has been rated at 8.67/10 (previous run: 7.69/10, +0.97)
```

### Expected behavior

No warning should be reported.

This may be controversial, but I wonder if it would make sense to silence the warnings in this case. just like they are inside of `__init__`. While `__new__` is used really rarely, it does not strike me as *wrong* to access private/protected members in there unless I am missing something. Would that really be a bad practice?

### pylint --version output

```
pylint 1.8.1,
astroid 1.6.0
Python 3.5.2 (default, Nov 23 2017, 16:37:01)
[GCC 5.4.0 20160609]
```

This is also somewhat related to #1801 as I noticed this error with that enum, but it turns out that the warning is also emitted for normal objects.

**Comment by PCManticore** [comment]
I'm not entirely sure that this would make sense in the general case. If we could infer that the left hand side of the assignment is an object generated by a call to that `__new__` then we could assume this is good enough and we are mutating the same object. But unless we can do that, ignoring by default messages emitted in `__new__` sounds a bit artificial, since some of them might be still valid (e.g one can use a different object to fetch a private attribute)

**Comment by exhuma** [comment]
> e.g one can use a different object to fetch a private attribute

I agree. Did not think of that one.

In the worst case, I think it is acceptable to append `# pylint: disable=<xyz>` to the violating line. This is in any case which should not happen *often*. If the implementation of this detection turns out to be a huge can of worms this could be seen as "not worth the effort" ;)

**Comment by limdor** [comment]
I'm having the same problem with the equal operator.
object.__eq__(self, other)
Is this the same case?
In case of Python 3 if I try object.__eq__(self, other: 'TypeHint') still complains.

**Comment by exhuma** [comment]
@limdor your case is a bit different as `__eq__` is an instance method. And there is something a bit more sinister at play: [Name Mangling](https://en.wikipedia.org/wiki/Name_mangling#Python)

In essence, the double-underscore-as-private is nothing more than syntactic sugar. The variable _is still accessible_ but it gets renamed. In the case of `__eq__` you don't know the class of `other` so blindly accessing a private attribute may cause an error. Even type hints won't help as they are not enforced at runtime (and I'm not sure that pylint honors type-hints at all).

So in that case, the pylint warning actually _does_ point to a potential bug in the code!

So when not being careful you might expect the following to return `True` (but it actually raises an exeption):

```
class MyClass:

    def __init__(self):
        self.__foo = 10

    def __eq__(self, other):
        return self.__foo == other.__foo


class MyOtherClass:

    def __init__(self):
        self.__foo = 10

    def __eq__(self, other):
        return self.__foo == other.__foo

if __name__ == '__main__':
    a = MyClass()
    b = MyOtherClass()
    print(a == b)
```

... however it raises an `AttributeError` with the following message: `'MyOtherClass' object has no attribute '_MyClass__foo'`

**Comment by limdor** [comment]
But what about with the attributes that have a single underscore? If we have this example with _foo instead of __foo, it would work but we would get the pylint warning.
Wouldn't be this the way to implement the `__eq__` operator of a custom class?

```
class MyClass:

    def __init__(self, value):
        self._foo = value

    def __eq__(self, other):
        if isinstance(other, self.__class__):
            return self._foo == other._foo
        else:
            return False


class MyOtherClass:

    def __init__(self, value):
        self._foo = value

    def __eq__(self, other):
        if isinstance(other, self.__class__):
            return self._foo == other._foo
        else:
            return False

if __name__ == '__main__':
    a = MyClass(10)
    b = MyOtherClass(10)
    c = MyClass(10)
    print(a == b)
    print(a == c)
```

**Comment by hippo91** [comment]
I agree with you @limdor. 
@PCManticore what do you think about it?

**Comment by PCManticore** [comment]
@hippo91 and @limdor Yes, this sounds like something that we can fix for single underscore attributes that are accessed in dunder methods.

**Comment by limdor** [comment]
Shouldn't be in general that a class method should be able to access the protected attributes of a parameter when it is known that is the same class?

**Comment by limdor** [comment]
Any update on this bug? Is it hard to implement? Some way to support?

**Comment by hippo91** [comment]
@limdor sadly i haven't got any time yet to work on this. I started this morning and seems quite easy.
I'll give you more informations ASAP.

**Comment by macdjord** [comment]
What about private variables of other instances of the same class? E.g.:
```python
class Foo:
    def __init__(self, child: 'Foo'):
        self.child: 'Foo' = child
        self._private: int = 1

    def do_thing(self):
        self.child._private += 1  # This should not raise a warning
```

**Comment by exhuma** [comment]
> What about private variables of other instances of the same class?

This feels a bit more controversial to me. When marking a variable as "private", you normally want to say that *no* external actor should be able to modify (or access) it. Think of two instances of `Car` for example. Should one instance be allowed to start the engine of the other?

If you expect that the attribute is accessed from the outside, then it might actually not be a good candidate for a "private" attribute.

My initial post was triggered by the `__new__` function, which is a bit special.

**Comment by macdjord** [comment]
That's not how private variables work in languages which enforce them. And it makes sense - if you have multiple instances of the same class, there needs to be some way for them to interact with each-other that other classes are not able to use, for instance to establish reciprocal parent-child relationships. Allowing them to access each other's private variables is one such method.

Indeed, if you look at how Python itself handles 'private' names - i.e. double-leading-underscore names - it does so in a way which prevents other classes from using them but no other instances of the same class.

**Comment by exhuma** [comment]
I decided to simplify the title a bit. I got confused myself every time I came across it myself. No idea what kind of carped I was smoking the day I wrote this... :sweat_smile: 

I also added a small paragraph in the beginning of the ticket, and it got me thinking: Could it be argued that the pattern of directly setting (d)under attributes inside `__new__` is bad in itself? By definition, `__init__` will be called after `__new__`, so my original post could actually set the attributes inside `__init__` instead. Which feels much more correct. And it might actually be _beneficial_ to have this issue reported by `pylint`.

This leaves the issue with methods like `__eq__`. You could argue that the current behvaiour of `pylint` is correct, making this whole issue moot. But being able to access protected/private attributes could make the code much easier in those cases. But this would only be okay if the accessed objet really _is_ of the correct class. So an  `isinstance` check is necessary before that.

The more I think about it, I beleive that the behaviour of `pylint` in its current state is correct. And if someone wants to access those attributes it cold be left up to the developer to add in a `# pylint: disable=protected-access` to the affected methods. That seems like a more sensible solution as situations like these _should_ be the exception.

After writing up this summary, I am okay with closing this ticket. My initial issue was with `__new__` and I can fix this by setting the args in `__init__`. Other "dunder" methods running into this issue should be an edge-case.

If the minds behind `pylint` agree with my assessment, feel free to close the issue. **Unless** you still think that methods like `__eq__` could benefit from this. I stand with my thoughts above though.

Consider the following:

```
# pylint: disable=missing-docstring, blacklisted-name, invalid-name
# pylint: enable=protected-access


class Hello:

    def __new__(cls):
        instance = object.__new__(cls)
        instance.__bar = 10  # <-- issue reported & bad practice
        return instance

    def __init__(self):
        print('__init__ called')
        self.__foo = 1  # <-- much better way of doing this

    def __eq__(self, other):
        return self.__foo == other.__foo  # <-- issue reported

    def get(self):
        return self.__foo


obj = Hello()
print(obj.get())
```

**Comment by exhuma** [comment]
> there needs to be some way for them to interact with each-other that other classes are not able to use, for instance to establish reciprocal parent-child relationships. Allowing them to access each other's private variables is one such method.

@macdjord Normally, this is what _protected_ variables are for. But those can only look "up" the inheritance tree. At least that's the intention of protected variables. Private variables are meant to be completely isolated in one instance and non-accessible from the outside.

Python is less strict with those rules and it lets you access them. But as I mentioned in the post I just wrote above, these are probably edge-cases, and it is arguable that these edge-cases could still be reported as errors by `pylint`, letting the developer add a `pylint: disable` line to the the affected block.

**Comment by macdjord** [comment]
@exhuma No, _protected_ variables (Python idiom: single leading underscore) are visible to a class and any class which inherits from it. Private variables (Python idiom: double leading underscore) are only visible within that class. I'm not aware of any strongly-typed language which actually has variables that are only visible within a specific _member_ of a class as you are suggesting.

And yes, I'm aware I can suppress these messages globally or by occurrence. *I shouldn't have to, because it's not an error.*

**Comment by hippo91** [comment]
Hi all and sorry for the delay.

If you have a look to the class `Issue1802` in the `pylint/test/functional/access_to_protected_members.py` file of the PR #2959 you could have an idea of what `pylint` will complain or not.
If you have some remarks about it don't hesitate to tell them right here.
Thanks

**Comment by limdor** [comment]
For me looks good

**Comment by ZeeD** [comment]
Sorry, but I'm trying to understand if my problem is related or not: I saw in this issue the "problem" with `__eq__(...)`; I'm facing something strange: if you look at the following snippet:

```
'module doc'


class MyClass:
    'class doc'

    def __init__(self) -> None:
        self._myprotected = 42

    def mymethod(self, other: 'MyClass') -> int:
        'method doc'
        return self._myprotected + other._myprotected

    def __eq__(self, other: object) -> bool:
        return self._myprotected == other._myprotected
```

pylint says:

```
PyLint: Executing command line: C:\Users\vito.detullio\Desktop\workspace\_scopa\.venv\Scripts\pylint.exe --output-format=text --msg-template='{C}:{line:3d},{column:2d}: {msg} ({symbol})' C:\Users\vito.detullio\Desktop\workspace\_scopa\src\protectedaccess.py
PyLint: The stdout of the command line is:
 ************* Module protectedaccess
W: 12,35: Access to a protected member _myprotected of a client class (protected-access)

------------------------------------------------------------------

Your code has been rated at 8.57/10 (previous run: 8.57/10, +0.00)
```

I'm trying to understand if there is a special case for __eq__ or something else.

**Comment by exhuma** [comment]
If I read this correctly, this is because you are accessing `_myprotected` on `other`. Which could be anything and is not guaranteed to have that attribute.

I don't know how smart `pylint` is in regards to type-checking, but maybe if you preface this with:

```
if not isinstance(other, MyClass):
    return False
```

it might work.

Or short-circuit it with:

```
def __eq__(self, other: object) -> bool:
        return isinstance(other, MyClass) and self._myprotected == other._myprotected
```

On a related note, shouldn't the type-hint of `other` be `Any` instead of `object`?

**Comment by PCManticore** [comment]
@ZeeD There is a special rule that allows this to properly work inside special methods, which is why you don't see it in `__eq__`.

Unfortunately, the solution proposed by @exhuma might not work since we don't take in account type guards done with `isinstance` but in the meantime, disabling the check in `mymethod` should do the trick (or exposing it as a public member if it's not so private)

**Comment by ZeeD** [comment]
Is there some plan to use the typing hints?
if not the general case, at least to check if a parameter of a method of a class is marked with the class itself (or a subclass). In this case I think it should be "acceptable" to use a protected method

**Comment by exhuma** [comment]
@ZeeD I find the combination of using `mypy` together with `pylint` works quite well for these cases. I personally don't see the point of adding support for type-hints into `pylint` because there is already a tool out there which does a really good job for that. Although, I'm not a pylint maintainer nor contributor, so I can't say what their position on this is.

I find that sometimes adding `pylint: disable...` and/or `type: ignore` comments is absolutely acceptable. I would recommend adding a small comment explaining *why* it was disabled though on the less trivial cases.

When starting with `pylint` and `mypy` I wanted to get my code "perfect", and I hated the `pylint: disable` and `type: ignore` comments to the core. But I've come to realise that it's absolutely acceptable, because, especially for the case of `pylint` there are moments where following it to the letter would make your code *less* readable/maintainable. And in those cases, just disabling the message in that block is a better choice than to try to "fix" it.
